x/wasm/IBC.md

```diff
@@ -178,96 +178,98 @@ type IBCCloseChannel struct {
-
-// how to handle error here? if we push it up the ibc stack and fail the transaction (normal handling),
-// the packet may be posted again and again. just log and ignore failures here? what does a failure even mean?
-// only realistic one I can imagine is OutOfGas (panic) to retry with higher gas limit.
-//
-// if there any point in returning a response, what does it mean?
-func IBCPacketAck(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, 
-    originalMsg []byte, result []byte) error {}
-
-// same question as for IBCPacketAck
-func IBCPacketDropped(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, 
-    originalMsg []byte, errorMsg string) error {}
-
-// do we need/want all this info?
-type IBCPacketInfo struct {
-	// local id for the Channel packet was sent/received on
-	ChannelID string
 
-    // sequence for the packet (will already be enforced in order if ORDERED channel)
-	Sequence uint64
-	
-    // Note: Timeout if guaranteed ONLY to be exceeded iff we are processing IBCPacketDropped
-    // otherwise, this is just interesting metadata
+// IBCContractCallbacks defines the methods for go-cosmwasm to interact with the wasm contract.
+// A mock contract would implement the interface to fully simulate a wasm contract's behaviour.
+type IBCContractCallbacks interface {
+  // Package livecycle
+
+  // OnIBCPacketReceive handles an incoming IBC package
+  OnIBCPacketReceive(hash []byte, params cosmwasm.Env, msg []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketReceiveResponse, uint64, error)
+  // OnIBCPacketAcknowledgement handles a IBC package execution on the counterparty chain
+  OnIBCPacketAcknowledgement(hash []byte, params cosmwasm.Env, originalData []byte, acknowledgement []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketAcknowledgementResponse, uint64, error)
+  // OnIBCPacketTimeout reverts state when the IBC package execution does not come in time
+  OnIBCPacketTimeout(hash []byte, params cosmwasm.Env, msg []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketTimeoutResponse, uint64, error)
+  // channel livecycle
+
+  // OnIBCChannelOpen does the protocol version negotiation during channel handshake phase
+  OnIBCChannelOpen(hash []byte, params cosmwasm.Env, order channeltypes.Order, version string, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCChannelOpenResponse, uint64, error)
+  // OnIBCChannelConnect callback when a IBC channel is established
+  OnIBCChannelConnect(hash []byte, params cosmwasm.Env, counterpartyPortID, counterpartyChannelID string, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCChannelConnectResponse, uint64, error)
```

Comment:
Same as above (wrap in stuct): `counterpartyPortID, counterpartyChannelID string` and please add the final `version` string here.. the second side may change it, so the original side must get that info at the end. And the `ordering` enum as well - so we have all info in one place when confirming the connection